zebra-state/src/service/finalized_state/zebra_db/shielded.rs

```diff
@@ -231,34 +234,131 @@ impl DiskWriteBatch {
         Ok(())
     }
 
-    /// Updates the supplied note commitment trees.
+    /// Update the supplied note commitment trees for the entire block,
+    /// using parallel `rayon` threads.
     ///
     /// If this method returns an error, it will be propagated,
     /// and the batch should not be written to the database.
     ///
     /// # Errors
     ///
     /// - Propagates any errors from updating note commitment trees
-    pub fn update_note_commitment_trees(
-        transaction: &Transaction,
+    pub fn update_note_commitment_trees_parallel(
+        block: &Arc<Block>,
         note_commitment_trees: &mut NoteCommitmentTrees,
     ) -> Result<(), BoxError> {
-        let sprout_nct = Arc::make_mut(&mut note_commitment_trees.sprout);
-        for sprout_note_commitment in transaction.sprout_note_commitments() {
-            sprout_nct.append(*sprout_note_commitment)?;
+        // Prepare arguments for parallel threads
+        let NoteCommitmentTrees {
+            sprout,
+            sapling,
+            orchard,
+        } = note_commitment_trees.clone();
+
+        let sprout_note_commitments: Vec<_> = block
+            .transactions
+            .iter()
+            .flat_map(|tx| tx.sprout_note_commitments())
+            .cloned()
+            .collect();
+        let sapling_note_commitments: Vec<_> = block
+            .transactions
+            .iter()
+            .flat_map(|tx| tx.sapling_note_commitments())
+            .cloned()
+            .collect();
+        let orchard_note_commitments: Vec<_> = block
+            .transactions
+            .iter()
+            .flat_map(|tx| tx.orchard_note_commitments())
+            .cloned()
+            .collect();
+
+        let mut sprout_result = None;
+        let mut sapling_result = None;
+        let mut orchard_result = None;
+
+        rayon::in_place_scope_fifo(|scope| {
+            if !sprout_note_commitments.is_empty() {
+                scope.spawn_fifo(|_scope| {
+                    sprout_result = Some(Self::update_sprout_note_commitment_tree(
+                        sprout,
+                        sprout_note_commitments,
+                    ));
+                });
+            }
+
+            if !sapling_note_commitments.is_empty() {
+                scope.spawn_fifo(|_scope| {
+                    sapling_result = Some(Self::update_sapling_note_commitment_tree(
+                        sapling,
+                        sapling_note_commitments,
+                    ));
+                });
+            }
+
+            if !orchard_note_commitments.is_empty() {
+                scope.spawn_fifo(|_scope| {
+                    orchard_result = Some(Self::update_orchard_note_commitment_tree(
+                        orchard,
+                        orchard_note_commitments,
+                    ));
+                });
+            }
+        });
```

Comment:
This is how I run sprout, sapling, and orchard note commitment tree updates in parallel in the finalized state.

Reply:
I moved this change to PR #4802, so @oxarbitrage can easily re-review this PR.